### ISSUE TITLE
Fix Realm-dynamic-simulator

### ISSUE THREAD

**Issue by shmidt** [realm/realm-cocoa#1327]
Can you please include dynamic frameworks example in sample code? 

Currently I do exactly as it says in docs and get error: ```ld: framework not found Realm-dynamic```

**Comment by jpsim** [comment]
Hi @shmidt, in attempting to follow our own installation steps, I realized that we need to make some changes to our distribution process to fully support installing the dynamic version of the Realm framework. Until we do this, the best way to install the dynamic version of Realm is to use [Carthage](https://github.com/Carthage/Carthage).

**Comment by jpsim** [comment]
Fixes have been made to support installing the dynamic Realm framework from our website download, but it won't be available until we cut our next release.

In the meantime, follow these steps to install the dynamic framework in an iOS 8 project:

1. Create two folders under the `ios/` directory of our download zip: `Realm-dynamic` and `Realm-dynamic-simulator`.
2. Drag `ios/Realm-dynamic.framework` into the `ios/Realm-dynamic` folder and `ios/Realm-dynamic-simulator.framework` into the `ios/Realm-dynamic-simulator` folder.
3. Rename the frameworks to `Realm.framework`.
4. Drag the `Realm-dynamic` & `Realm-dynamic-simulator` folders from
   the `ios/` directory to the File Navigator of your Xcode project.
   Make sure **Copy items if needed** is selected and click **Finish**.
5. In your app target's **Build Phases** tab, click the + button in the top
   left of the panel and select **New Copy Files Phase**.
   Rename this new phase to **Copy Frameworks**, set the **Destination** to
   **Frameworks**, and add `Realm-dynamic/Realm.framework`.

**Comment by shmidt** [comment]
@jpsim Thank you.
I did exactly as it was written by you,
but I still have this issue.

Should I remove one of the frameworks from the linked section, probably? 
![screen shot 2015-01-22 at 17 56 53](https://cloud.githubusercontent.com/assets/1472370/5857911/4a8b119e-a260-11e4-8b0a-f0e2f380dda3.png)

**Comment by shmidt** [comment]
Oh, and I see that error message has changed:
```swift
ld: warning: ignoring file /Users/schmidt/Documents/iOS/VoiceMemo/VoiceMemo/Realm/Realm-dynamic/Realm.framework/Realm, missing required architecture x86_64 in file /Users/schmidt/Documents/iOS/VoiceMemo/VoiceMemo/Realm/Realm-dynamic/Realm.framework/Realm (2 slices)
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_RLMArray", referenced from:
      __TMaCSo8RLMArray in RLMSupport.o
  "_OBJC_CLASS_$_RLMObject", referenced from:
      __TFFE9VoiceMemoCSo8RLMArray8generateFS0_FT_GVSs11GeneratorOfCSo9RLMObject_U_FT_GSqS2__ in RLMSupport.o
      __TMaCSo9RLMObject in RLMSupport.o
      __TFFE9VoiceMemoCSo10RLMResults8generateFS0_FT_GVSs11GeneratorOfCSo9RLMObject_U_FT_GSqS2__ in RLMSupport.o
  "_OBJC_CLASS_$_RLMResults", referenced from:
      __TMaCSo10RLMResults in RLMSupport.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by shmidt** [comment]
I see that framework search paths has records to both ```Realm-dynamic``` and ```Realm-dynamic-simulator```, maybe that is the problem.

**Comment by jpsim** [comment]
I'm looking into this, @shmidt. Bear with me a bit.

**Comment by jpsim** [comment]
So it turns out that dragging from the sidebar into the "Copy Files" phase is different than adding the file by clicking the "+" icon. So try removing the file from the "Copy Files"/"Copy Frameworks" phase and re-adding it by clicking "+".

It feels a bit silly, but I recorded a video going through these steps from scratch with our latest release (0.90.0): http://static.realm.io/videos/realm-dynamic-test.mp4

Who knows, maybe it'll help you figure out what you're doing differently. Keep us posted!

**Comment by shmidt** [comment]
@jpsim Thanks for the video. I do the same, except I have Swift project.

**Comment by shmidt** [comment]
@jpsim I followed your video step by step, to create exactly the same project, but result is the same:
```
ld: warning: ignoring file RealmDynamicTest/Realm-dynamic/Realm.framework/Realm, missing required architecture x86_64 in file /Users/schmidt/Documents/iOS/RealmDynamicTest/Realm-dynamic/Realm.framework/Realm (2 slices)
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_RLMRealm", referenced from:
      objc-class-ref in ViewController.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Can't understand what the problem this. 
My Xcode version is up to date: Version 6.1.1 (6A2008a).
Probably re-installing Xcode can fix it?

**Comment by shmidt** [comment]
And this is how architectures look like:
![screen shot 2015-01-27 at 18 09 26](https://cloud.githubusercontent.com/assets/1472370/5920720/bc762260-a64f-11e4-9a3f-9a51a604509e.png)

**Comment by shmidt** [comment]
Can you please try to build my sample project?
https://dl.dropboxusercontent.com/u/4110416/RealmDynamicTest.zip
If it will compile without errors then I will know that problem is in Xcode.

**Comment by pizthewiz** [comment]
Yup, I have the same linker issue - this seems to be caused by linking to both versions of the framework. For example, if I follow the normal instructions or the video above and bring in both versions of the framework, I can build for the device but not the simulator. If I then remove the device version of the framework and clean up my project's `FRAMEWORK_SEARCH_PATHS`, I can then build for the simulator.

A warning is displayed at link time saying that it is skipping the other version, but it seems to be using it still.

**Comment by pizthewiz** [comment]
I see the same issues in @shmidt's sample project when building for the simulator as well.

If we look at the build log for the linking step we'll see:
```

-F/Users/xyzzy/Downloads/RealmDynamicTest/Realm-dynamic 
-F/Users/xyzzy/Downloads/RealmDynamicTest/Realm-dynamic-simulator

-framework Realm -framework Realm
```

and a warning:
```
ld: warning: ignoring file /Users/xyzzy/Downloads/RealmDynamicTest/Realm-dynamic/Realm.framework/Realm, missing required architecture x86_64 in file /Users/xyzzy/Downloads/RealmDynamicTest/Realm-dynamic/Realm.framework/Realm (2 slices)
```

It would seem that it is *not* actually ignoring the file.

**Comment by pizthewiz** [comment]
Just re-read the commentary in #1163, doesn't seem that there is an easy fix for this…

**Comment by jpsim** [comment]
Hi all, I reopened this as an issue and we'll fix as soon as possible. The dynamic framework is currently working for devices, but not for the simulator. It's an architecture issue.

In the meantime, the best way to use the dynamic Realm framework is through Carthage. We understand this may be inconvenient for some, so we're working on making the `Realm-dynamic-simulator` framework behave properly.

**Comment by shmidt** [comment]
@jpsim Good to hear. Thank you very much!

**Comment by pizthewiz** [comment]
I think this isn't actually an issue with the Realm-dynamic-simulator framework, but an issue of linking to two frameworks of the same name - with some renaming, I suspect one could cause the exact opposite situation, where linking for device fails but for simulator does not.

**Comment by jpsim** [comment]
> issue of linking to two frameworks of the same name

Possibly. Unfortunately we can't rename the frameworks (that's what we were doing initially). We could also go the Carthage route and require that users add a build script to strip architectures on submission to the App Store.

**Comment by pizthewiz** [comment]
A run script build phase to strip the architectures does not seem unreasonable to me as it is not uncommon to have a separate scheme to target App Store submission.

**Comment by pizthewiz** [comment]
Is there any update on this? Without a usable dynamic framework option, linking against the static Realm framework in an iOS app uses an embedded framework and app extension results in duplicate symbols and persistence exceptions as #1408 reported.

**Comment by pizthewiz** [comment]
A run script build phase that strips the unneeded architectures seems like a very good option until (or if) the official Xcode tooling supports it.

**Comment by jpsim** [comment]
We're still working on improving this situation.

If you need to use the dynamic framework version of Realm today, I'd recommend that you either build from source, use Carthage or use CocoaPods with the `use_frameworks!` flag.

**Comment by pizthewiz** [comment]
Carthage it is, thanks!

Would it make sense to add the dynamic framework issue to the [Current Limitations](http://realm.io/docs/cocoa/0.91.1/#current-limitations) list?
